pop-api/src/error.rs

```diff
@@ -0,0 +1,255 @@
+use crate::assets::use_cases::fungibles::{convert_to_fungibles_error, FungiblesError};
+use ink::env::chain_extension::FromStatusCode;
+use scale::{Decode, Encode};
+use PopApiError::*;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+#[repr(u8)]
+pub enum PopApiError {
+	/// Some unknown error occurred. Go to the Pop API docs section `Pop API error`.
+	Other {
+		// Index within the `DispatchError`
+		dispatch_error_index: u8,
+		// Index within the `DispatchError` variant.
+		error_index: u8,
+		// Index for further nesting, e.g. pallet error.
+		error: u8,
+	} = 0,
+	/// Failed to lookup some data.
+	CannotLookup = 1,
+	/// A bad origin.
+	BadOrigin = 2,
+	/// A custom error in a module.
+	Module {
+		index: u8,
+		error: u8,
+	} = 3,
+	/// At least one consumer is remaining so the account cannot be destroyed.
+	ConsumerRemaining = 4,
+	/// There are no providers so the account cannot be created.
+	NoProviders = 5,
+	/// There are too many consumers so the account cannot be created.
+	TooManyConsumers = 6,
+	/// An error to do with tokens.
+	Token(TokenError) = 7,
+	/// An arithmetic error.
+	Arithmetic(ArithmeticError) = 8,
+	/// The number of transactional layers has been reached, or we are not in a transactional
+	/// layer.
+	Transactional(TransactionalError) = 9,
+	/// Resources exhausted, e.g. attempt to read/write data which is too large to manipulate.
+	Exhausted = 10,
+	/// The state is corrupt; this is generally not going to fix itself.
+	Corruption = 11,
+	/// Some resource (e.g. a preimage) is unavailable right now. This might fix itself later.
+	Unavailable = 12,
+	/// Root origin is not allowed.
+	RootNotAllowed = 13,
+	// TODO: make generic and add docs.
+	UseCaseError(FungiblesError) = 254,
+	DecodingFailed = 255,
+}
+
+impl FromStatusCode for PopApiError {
+	fn from_status_code(status_code: u32) -> core::result::Result<(), Self> {
+		match status_code {
+			0 => Ok(()),
+			_ => Err(convert_to_pop_api_error(status_code)),
+		}
+	}
+}
+
+// `pub` because it is used in `runtime/devnet/src/extensions/tests/mod.rs`'s test:
+// `dispatch_error_to_status_code_to_pop_api_error_works`
+//
+// This function converts a given `status_code` (u32) into a `PopApiError`. First it encodes the
+// status code into a 4-byte array and checks for unknown nested errors. If decoding into
+// `PopApiError` fails (e.g. a breaking change in the `DispatchError`), it handles the error by
+// converting it to the `Other` variant by shifting each byte one position forward (the last byte is
+// not used for anything)and setting the first byte to 0. If decoding succeeds, it checks if the
+// error is of the `Module` variant and performs any necessary conversion based on the use case.
+pub fn convert_to_pop_api_error(status_code: u32) -> PopApiError {
+	let mut encoded: [u8; 4] =
+		status_code.encode().try_into().expect("qid u32 always encodes to 4 bytes");
```

Comment:
Is it perhaps more efficient to just get the bytes from the u32 rather than the scale encoded version? u32 encoding (endianness) another consideration.

Reply:
Great suggestion, thanks!